src/avram/upsert.cr

```diff
@@ -0,0 +1,104 @@
+  #
+  # **Note:** In almost _every_ case the `upsert_lookup_columns` should have a **unique index** defined
+  # in the database to ensure no conflicting records are created, even from outside Avram.
+  #
+  # ## Full Example
+  #
+  # ```
+  # class User < Model
+  #   table do
+  #     column name : String
+  #     column email : String # This column has a unique index
+  #   end
+  # end
+  #
+  # class SaveUser < User::SaveOperation
+  #   # Can be one or more columns. In this case we choose just :email
+  #   upsert_lookup_columns :email
+  # end
+  #
+  # # Will create a new row in the database since no row with
+  # # `email: "bob@example.com"` exists yet
+  # SaveUser.upsert!(name: "Bobby", email: "bob@example.com")
+  #
+  # # Will update the name on the row we just created since the email is
+  # # the same as one in the database
+  # SaveUser.upsert!(name: "Bob", email: "bob@example.com")
+  # ```
+  #
+  # ## Difference between `upsert` and `upsert!`
+  #
+  # There is an `upsert` and `upsert!` that work similarly to `create` and `create!`.
+  # `upsert!` will raise an error if the operation is invalid. Whereas `upsert`
+  # will yield the operation and the new record if the operation is valid, or
+  # the operation and `nil` if it is invalid.
+  #
+  # ```
+  # # Will raise because the name is blank
+  # SaveUser.upsert!(name: "", email: "bob@example.com")
+  #
+  # # Operation is invalid because name is blank
+  # SaveUser.upsert(name: "", email: "bob@example.com") do |operation, user|
+  #   # `user` is `nil` because the operation is invalid.
+  #   # If the `name` was valid `user` would be the newly created user
+  # end
+  # ````
+  macro upsert_lookup_columns(*attribute_names)
+    def self.upsert!(*args, **named_args) : T
+      operation = new(*args, **named_args)
+      existing_record = find_existing_unique_record(operation)
+
+      if existing_record
+        operation.record = existing_record
+      end
+
+      operation.save!
+    end
+
+    def self.upsert(*args, **named_args)
+      operation = new(*args, **named_args)
+      existing_record = find_existing_unique_record(operation)
+
+      if existing_record
+        operation.record = existing_record
+      end
+
+      operation.save
+      yield operation, operation.record
+    end
+
+    def self.find_existing_unique_record(operation) : T?
+      T::BaseQuery.new
+        {% for attribute in attribute_names %}
+          .{{ attribute.id }}.nilable_eq(operation.{{ attribute.id }}.value)
+        {% end %}
+        .first?
+    end
+  end
+
+  # :nodoc:
+  macro included
+    {% for method in ["upsert", "upsert!"] %}
+      # Performs a create or update depending on if there is a conflicting row in the database.
+      #
+      # See `Avram::Upsert.upsert_lookup_columns` for full documentation and examples.
+      def self.{{ method.id }}(*args, **named_args)
+        \{% raise "Please call 'upsert_lookup_columns' on #{@type} before using '{{ method.id }}'" %}
```

Comment:
Just thinking like a newbie here, this might lead me to believe I need to do something like:

```crystal
SaveUser.upsert_lookup_columns
SaveUser.upsert!(...)
```

Maybe we can word it like "Please include the `upsert_lookup_columns` macro in #{@type} before using '{{ methods.id }}'" ?

Reply:
Good idea! THat makes a lot of sense